### PR TITLE
mdatagen: support custom package name

### LINKED ISSUE #11231: Allow `mdatagen` to generate packages with names other than `metadata`
**Is your feature request related to a problem? Please describe.**
In contrib, receivers will soon need to make shifts in the metrics they produce to align with Semantic Conventions. Depending on what exact changes are necessary, this is really hard to do with only one `metadata` package, especially if you want to not break existing usage.

**Describe the solution you'd like**
I would like `mdatagen` to be able to generate packages with names other than `metadata`. This would allow receivers making the semconv transition to have an entirely separate yaml file and generated package with the semconv version of the produced metrics, and for example control the produced set of metrics via featuregate.

**Describe alternatives you've considered**
I already tried to do this all in one `metadata.yaml` and `metadata` package, but it was messy and in some cases had conflicts that were impossible to reconcile.

### PR DESCRIPTION
**Pull request by braydonk**
#### Description
This PR adds support for a new ~~`--package_name` flag~~ `generated_package_name` config field to allow mdatagen to generate packages with names other than `metadata`.

#### Link to tracking issue
Fixes #11231 

#### Testing
* Unit tests
* `go install`ing this branch and using it in a test scenario with two yaml files and generating two packages.

### CODE REVIEW

File: cmd/mdatagen/README.md [hunk 1]
```diff
@@ -55,6 +55,27 @@ Below are some more examples that can be used for reference:
 
 You can run `cd cmd/mdatagen && $(GOCMD) install .` to install the `mdatagen` tool in `GOBIN` and then run `mdatagen metadata.yaml` to generate documentation for a specific component or you can run `make generate` to generate documentation for all components.
 
+### Generate a package with an alternate name
+
+By default, `mdatagen` will generate a package called `metadata` in the `internal` directory. If you want to generate a package with a different name, you can use the `--package_name` flag to provide an alternate name.
+
+```go
+//go:generate mdatagen --package_name custom custom.yaml
```
Comment (bogdandrutu) [inline cmd/mdatagen/README.md:63]: Why not have it in the yaml config?

Comment (braydonk) [reply cmd/mdatagen/README.md:63]: Good idea, I'm happy to change it to that instead.

Comment (bogdandrutu) [reply cmd/mdatagen/README.md:63]: Let's try to keep the entire config in the file and avoid mixing flags and file if possible.

Comment (braydonk) [reply cmd/mdatagen/README.md]: Done in [d3f5e4e](https://github.com/open-telemetry/opentelemetry-collector/pull/11232/commits/d3f5e4eddc6dd5f02f108876d92f7e7544238c5f)

File: cmd/mdatagen/generated_package_test.go [hunk 2]
```diff
@@ -3,9 +3,8 @@
 package main
 
 import (
-	"testing"
-
 	"go.uber.org/goleak"
+	"testing"
```
Comment (codeboten) [inline cmd/mdatagen/generated_package_test.go:7]: This change should be reverted

File: cmd/mdatagen/internal/samplereceiver/generated_package_test.go [hunk 3]
```diff
@@ -3,9 +3,8 @@
 package samplereceiver
 
 import (
-	"testing"
-
 	"go.uber.org/goleak"
+	"testing"
```
Comment (codeboten) [inline cmd/mdatagen/internal/samplereceiver/generated_package_test.go:7]: Same comment here

File: cmd/mdatagen/internal/samplereceiver/generated_component_test.go [hunk 4]
```diff
@@ -8,7 +8,6 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/require"
-
 	"go.opentelemetry.io/collector/component"
```
Comment (codeboten) [inline cmd/mdatagen/internal/samplereceiver/generated_component_test.go:11]: Same comment here

File: cmd/mdatagen/internal/samplereceiver/internal/metadata/generated_config_test.go [hunk 5]
```diff
@@ -9,7 +9,6 @@ import (
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-cmp/cmp/cmpopts"
 	"github.com/stretchr/testify/require"
-
```
Comment (codeboten) [inline cmd/mdatagen/internal/samplereceiver/internal/metadata/generated_config_test.go:12]: ditto

File: cmd/mdatagen/internal/samplereceiver/internal/metadata/generated_metrics_test.go [hunk 6]
```diff
@@ -6,12 +6,11 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
-	"go.uber.org/zap"
-	"go.uber.org/zap/zaptest/observer"
-
```
Comment (codeboten) [inline cmd/mdatagen/internal/samplereceiver/internal/metadata/generated_metrics_test.go:11]: ditto

**Review by codeboten — COMMENTED** [review]
Overall this looks pretty good, just a couple of comments, it looks like `make gofmt` needs to be run to sort out the import statements. Also please add a note in the schema file: https://github.com/open-telemetry/opentelemetry-collector/blob/main/cmd/mdatagen/metadata-schema.yaml

**Comment by braydonk** [comment]
> Also please add a note in the schema file

Done in [d6c50bf](https://github.com/open-telemetry/opentelemetry-collector/pull/11232/commits/d6c50bf112ca0e00f5a78ad4de9c36e2cf62587d)

**Review by codeboten — APPROVED** [review]
Thanks for the update @braydonk, looks good now, can merge it once the conflict has been resolved

**Review by dmitryax — APPROVED** [review]

**Pull request merged**